### ISSUE TITLE
MQTT / X.509 authentication broken

### ISSUE THREAD

**Issue by echodelt** [thingsboard/thingsboard#1355]
Hi,

The MQTT / X.509 device authentication feature seems to be broken since the release of the Thingsboard 2.2.0 version.

I could previously use it with TB 2.0 and TB 2.1.3, pushing data with a mosquitto client.

Now with the same server MQTT / SSL configuration and the same server and device certificates, I get this error message in the log file : 

`[nioEventLoopGroup-5-5] INFO  o.t.s.t.mqtt.MqttTransportHandler - [localhost:41928] Invalid message received`

(no more details available in the log file)

Here are my server settings :

```
  # Local MQTT transport parameters
  mqtt:
    enabled: "${MQTT_ENABLED:true}"
    bind_address: "${MQTT_BIND_ADDRESS:0.0.0.0}"
    bind_port: "${MQTT_BIND_PORT:8883}"
    timeout: "${MQTT_TIMEOUT:10000}"
    netty:
      leak_detector_level: "${NETTY_LEAK_DETECTOR_LVL:DISABLED}"
      boss_group_thread_count: "${NETTY_BOSS_GROUP_THREADS:1}"
      worker_group_thread_count: "${NETTY_WORKER_GROUP_THREADS:12}"
      max_payload_size: "${NETTY_MAX_PAYLOAD_SIZE:65536}"

  ssl:
    enabled: "${MQTT_SSL_ENABLED:true}"
    protocol: "${MQTT_SSL_PROTOCOL:TLSv1.2}"
    key_store: "${MQTT_SSL_KEY_STORE:mqtt_server.tb1.jks}"
    key_store_password: "${MQTT_SSL_KEY_STORE_PASSWORD:my_password}"
    key_password: "${MQTT_SSL_KEY_PASSWORD:my_password}"
    key_store_type: "${MQTT_SSL_KEY_STORE_TYPE:JKS}"
```
And data is pushed with :

`mosquitto_pub -h "tb1" -p 8883  --tls-version tlsv1.2 -t "v1/devices/me/telemetry" --cafile mqtt_server.tb1.pub.pem --key mqttclient.key --cert mqttclient.pem -m '{"temperature":12.3, "humidity":45.6}' `

Has anyone else encountered a problem with X.509 authentication using TB 2.2 ?

Greetings,

Thanks

**Comment by ashvayka** [comment]
Fix will be available in v2.2.1 and is already in master branch.

**Comment by souhaiebtar** [comment]
@echodelt  from where do we get `mqttclient.key` file, because i can't find it

**Comment by mp-loki** [comment]
@souhaiebtar you need to run client.keygen.sh script to get it

**Comment by souhaiebtar** [comment]
@mp-loki i already run it, i get many files `.pem` `.no.pass.pem` `.p12` but the script doesn't generate a `.key` file

**Comment by souhaiebtar** [comment]
@mp-loki maybe a new version submitted which doesn't generate `.key` file, i tried even to go back to older commit but i can't find  a script that generate a `.key`

**Comment by mp-loki** [comment]
According to mosquitto_pub man:

` --key : client private key for authentication, if required by server.`

Normally the server should not require the client private key (that's the reason why it's private), and ThingsBoard server does not. If this is still the case, you can try to export client private key using keytool from mqttclient.jks

**Comment by souhaiebtar** [comment]
@mp-loki thanks for everything, i already try it without it but without success, i will try to `export client private key using keytool`

**Comment by echodelt** [comment]
Hi,
Here are some details on the way I generate and use the devices certificates for X.509 authentication. 

At first, you need, as you mentioned it, to generate the SERVER certificate and enable MQTT over SSL as explained [here](https://thingsboard.io/docs/user-guide/mqtt-over-ssl/).

For this, you'll have to edit a `keygen.properties` file and run the `server.keygen.sh` script. The script will generate several files including a `.pem.pub` one. This `.pem.pub` file (or the `.cer` one) will have to be communicated later to the clients.

Of course, don't forget to reconfigure and restart Thingsboard in order to enable MQTT over SSL. At this stage, you should be able to use MQTT and "one-way" SSL with something like :

`mosquitto_pub -h "tb.test.com" -p 8883 --tls-version tlsv1.2 -t "v1/devices/me/telemetry" -u 'qwertyuiop' --cafile mqtt_server.tb.test.com.pub.pem -m '{"temperature":12.3, "humidity":45.6}' `

Now if you want to publish data using X.509 authentication, move, for example, on an other Linux station and generate, on it, the DEVICE authentication key pair using, for example, this command :

`openssl req -x509 -sha256 -nodes -days 999 -newkey rsa:2048 -keyout mqttclient.key -out mqttclient.pem -subj "/C=FR/O=TA/OU=TA/CN=tb.test.com"`

This is at this stage that the the `mqttclient.key` and `mqttclient.pem` are produced.

Then copy the `mqttclient.pem` file content and paste it in the 'Device Credentials' popup, in the  device configuration window.

If everything's OK, you should be able to publish data using :

`mosquitto_pub -h "tb.test.com" -p 8883 --tls-version tlsv1.2 -t "v1/devices/me/telemetry" --cafile mqtt_server.tb.test.com.pub.pem --key mqttclient.key --cert mqttclient.pem -m '{"temperature":12.3, "humidity":45.6}' `

As you can see, the devices authentication key pairs can be generated on other Linux stations than the one who hosts the ThingsBoard server. And of course stored outside of the JKS keystore (or something similar) which exists on this one.

**Comment by souhaiebtar** [comment]
@echodelt thanks for everything

**Comment by paolofrs** [comment]
@echodelt thanks for your explenation. Unfortunately something went wrong on my side. 

I am connecting using one-way-ssl (only the server authenticate via SSL) with a mosquitto client. When I try to publish some messages the ThingsBoard app throws and error saying it does not find the `/some/directory/keys/mqttserv.jks`

The _thingsboard.yml_ file has  

`key_store: "${MQTT_SSL_KEY_STORE:/some/directory/keys/mqtt_server.jks}"`  

and the file in `/some/directory/keys/mqttserver.jks` exist and it has correct permissions (`chwon 400 /some/directory/keys/mqtt_server.jks` and `chmod thingsboard:thingsboard /some/directory/keys/mqtt_server.jks`)

 **Where should I place `mqttserver.jks` in order to be found by the app at run time?**

**Comment by echodelt** [comment]
Hi @paolofrs ,

I's still running TB 2.3.1. When I setup it I had to :

- put the `tb-server.jks` file in the `/etc/thingsboard/conf/` directory
- `chmod 400` and `chown thingsboard:thingsboard` it
- use a relative reference to this file in `thingsboard.yml` :
  `key_store: "${MQTT_SSL_KEY_STORE:tb-server.jks}"`

If I remember well, I also tried at the time to use an absolute path but without success. And ended to use a relative reference instead.
